### DISCUSSION TITLE
I have two GLB files and I would like to apply the animation defined in file A to  file B.

### DISCUSSION THREAD

**Discussion by atul161** [donmccurdy/glTF-Transform#730] [Q&A] [answered]
I have two GLB files and I would like to apply the animation defined in file A to  file B. Is it possible?

```
    const animationDocument: Document = await readGlb('./animation/animation.glb')
    // @ts-ignore
    let animations: Animation[] = animationDocument.getRoot().listAnimations();
    let animationChannel: AnimationChannel[];
    let animationSampler: AnimationSampler[];
    for(let i = 0; i < animations?.length ; i++){
       animationChannel = animations[i]?.listChannels()
       animationSampler = animations[i]?.listSamplers()
    }
    const avtarDocument: Document = await readGlb('./assets/atul.glb')
    // @ts-ignore
    avtarDocument.createAnimation(animations[0].getName()).addChannel(animationChannel[0]).addSampler(animationSampler[0])

```

This is giving error Cannot connect disconnected graphs.

Can anyone please help how to solve this problem.

**Comment by donmccurdy** [comment, marked as answer]
This is possible, but I don't have an "out of the box" code snippet to recommend. Being able to combine the files at all depends on some assumptions about the content of the files, such as:

1. Do the animations use skinning/armatures, or simple objects with keyframe position/rotation/scale?
2. If skinning, do both files contain the full character? just skeletons? if skeletons, are the skeletons _identical_ (same joints, joint names, joint default pose, bind matrices, ...)?

Based on the assumption that you have two characters with the same exact skeleton, the steps will be as discussed in https://github.com/donmccurdy/glTF-Transform/discussions/665#discussioncomment-3484330. Copying here:

1. merge the glTF files containing the skinned mesh and the animation(s) e.g. with `gltf-transform merge`
2. find the primary [Skin](https://gltf-transform.donmccurdy.com/classes/core.skin.html) (the one you want to keep), and gather a list of its joints (`skin.listJoints()`) 
3. for each [AnimationChannel](https://gltf-transform.donmccurdy.com/classes/core.animationchannel.html), replace its target node (currently from the animation-only glTF) with the target node of the same name from the primary Skin
4. dispose all unused Skins, Scenes, and Nodes

***

The "Cannot connect disconnected graphs." error occurs because the first step is missing, the Documents loaded from each file must first be merged into one Document, then you can continue with consolidating the animations and skins.

**Reply by atul161** [reply]
Thanks ! You are my saviour!!

**Reply by atul161** [reply]
Hii @donmccurdy ! ,

Thanks for answering my question. I Just need to ask one more question regarding an animation.
Currently i have an animation glb which have 5 types of animations. ie running , walking , victory and etc.

According to your answer i have successuffy merged my avtaar glb into animation glb. Here in animation glb i have only animations and its nodes matching to the same node of avtaar glb.

When i created a new glb after merging. The first animation is autoplayed in new glb when i render in gltf viewer. Is there any option or way so that i can autoplay 2 animation in glb. 


Thanks!!

**Reply by donmccurdy** [reply]
Glad things are working so far!

A glTF file does not contain anything that tells a viewer which animation(s) autoplay, but the convention for viewers is to play the first animation by default. You could overwrite the first animation to contain all the channels and samplers from the two you want to play (which doesn't duplicate any keyframe data), and then viewers would use that. 

This may not work if the two animations target the same bones in the character. That gets into weighted blending, and will be more complicated that just merging appending the channels and samplers to another animation.
